compiler/codegen/CodeGenPrep.cpp

```diff
@@ -177,6 +179,128 @@ OMR::CodeGenerator::lowerTreesPreChildrenVisit(TR::Node * parent, TR::TreeTop *
+ * other threads.
+ *
+ * @param comp pointer to the compilation object
+ * @param node the node being lowered
+ * @param tt the TreeTop anchoring the node
+ */
+static void
+lowerNewValue(TR::Compilation *comp, TR::Node *node, TR::TreeTop *tt)
+   {
+   auto* prevTreeTop = tt->getPrevTreeTop();
+
+   // remove tt from the TreeTop List
+   prevTreeTop->setNextTreeTop(NULL);
+
+   // Iterate over every child that sets a field. The first
+   // child is skipped as it's only used to specify the type
+   // of the value being constructed. The second child is
+   // the one that actually initializes the first field.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      auto* ttNode = TR::Node::create(TR::treetop, 1);
+      ttNode->setFirst(node->getChild(i));
+      prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+      prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+      }
+
+   // Create new TreeTop to anchor the new. The TreeTop for
+   // the newValue cannot be reused because it is still used
+   // just after lowering the node.
+   auto* newNode = TR::Node::createWithSymRef(TR::New, 1, node->getSymbolReference());
+   newNode->setFirst(node->getChild(0));
+   auto* ttNode = TR::Node::create(TR::treetop, 1);
+   ttNode->setAndIncChild(0, newNode);
+   prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+   prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+
+   // Get the class and layout of the value type
+   auto* valueClass = static_cast<TR_OpaqueClassBlock *>(static_cast<TR::StaticSymbol *>(newNode->getSymbol())->getStaticAddress());
+   const TR::TypeLayout* typeLayout = comp->typeLayout(valueClass);
+
+   // Iterate over every child that sets a field and generated
+   // a store for the value.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      TR::IL il;
+      auto storeOpCode = il.opCodeForIndirectStore(node->getChild(i)->getDataType());
+      const TR::TypeLayoutEntry fieldEntry = typeLayout->entry(i - 1);
+      const bool isVolatile = false; // fields of newvalue allocated objects are,
+      const bool isPrivate = false;  // by definition, non-volatile, non-private,
+      const bool isFinal = true;     // and final
+      auto* symref = comp->getSymRefTab()->findOrFabricateShadowSymbol(valueClass,
+                                                                       fieldEntry._datatype,
+                                                                       fieldEntry._offset,
+                                                                       isVolatile,
+                                                                       isPrivate,
+                                                                       isFinal,
+                                                                       fieldEntry._fieldname,
+                                                                       TR::DataType::getPrefix(fieldEntry._datatype)
```

Comment:
I'm not sure if this is correct because the "prefix" doesn't strictly match the signature of the type. Should the signature of the type be added to `TypeLayoutEntry`?

Reply:
Yeah I think we would want to keep the proper signature in a downstream project like OpenJ9 - question is - is the prefix 'good enough' as a default implementation - probably. If the proper signature version could be pulled up to the OMR level I thin it would be preferable.

Reply:
I know the prefix doesn't quite match the signature (even in OpenJ9) so I've added a field for the signature in the TypeLayout class. That way, downstream projects can set whatever signature is appropriate for their language.